File: 프로그래머스/2/12952. N－Queen/N－Queen.js
function solution(n) {
    let answer = 0;
    const width = Array(n).fill(false);
    const diagonal1 = Array(n * 2).fill(false);
    const diagonal2 = Array(n * 2).fill(false);
    
    function dfs(y) {
        if(y === n) {
            answer += 1;
        } else {
            for(let i = 0; i < n; i++) {
                if(width[i] || diagonal1[i + y] || diagonal2[i - y + n]) {
                    continue;
                }
                width[i] = diagonal1[i + y] = diagonal2[i - y + n] = true;
                dfs(y + 1);
                width[i] = diagonal1[i + y] = diagonal2[i - y + n] = false;
            }
        }
    }
    dfs(0);
    return answer;
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    answer = 0
    line = [False] * n # 
    diagonal1 = [False] * 2 * n # 우상향 대각선
    diagonal2 = [False] * 2 * n # 우하향 대각선
    
    def backtrack(y):
        nonlocal answer
        if y == n:
            answer += 1
            return
        for i in range(n):
            if line[i] or diagonal1[i + y] or diagonal2[i + (n - y)]:
                continue
            line[i] = diagonal1[i + y] = diagonal2[i + (n - y)] = True
            backtrack(y + 1)
            line[i] = diagonal1[i + y] = diagonal2[i + (n - y)] = False
    
    backtrack(0)    
    return answer